File: algorithms/[82]删除排序链表中的重复元素 II/solution.ts
// topics = ["链表"]

export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function deleteDuplicates(head: ListNode | null): ListNode | null {
  const dummy = new ListNode();
  dummy.next = head;

  let pre = dummy;
  let val: number | undefined;
  // 遍历非尾节点
  while (head && head.next) {
    if (head.val == head.next.val || head.val == val) {
      val = head.val;
      head.val = head.next.val;
      head.next = head.next.next;
    } else {
      pre = head;
      head = head.next;
    }
  }

  // 单独判断尾节点
  if (head?.val == val) {
    pre.next = null;
  }

  return dummy.next;
}

File: algorithms/[82]删除排序链表中的重复元素 II/solution.py
# topics = ["链表"]

from __future__ import annotations
from typing import Union


class ListNode:
    def __init__(self, val: int = 0, next: Union[ListNode, None] = None):
        self.val = val
        self.next: Union[ListNode, None] = next


class Solution:
    def deleteDuplicates(self, head: ListNode) -> Union[ListNode, None]:
        if not head:
            return head

        dummy = ListNode(0, head)
        cur = dummy

        while cur.next and cur.next.next:
            if cur.next.val == cur.next.next.val:
                x = cur.next.val
                while cur.next and cur.next.val == x:
                    cur.next = cur.next.next
            else:
                cur = cur.next

        return dummy.next
